dislib/decomposition/tsqr/base.py

```diff
@@ -71,293 +67,693 @@ def tsqr(a: Array, n_reduction=2, mode="complete", indexes=None):
+                                          q_blocks,
+                                          q_blocks_2,
+                                          r_blocks,
+                                          last_block_irregular,
+                                          irregular_shape_to_use,
+                                          "complete")
+                    small_q = Array(q_blocks, top_left_shape=(shape_to_use,
+                                                              shape_to_use),
+                                    reg_shape=(shape_to_use, shape_to_use),
+                                    shape=(shape_to_use * auxiliar_rs,
+                                           shape_to_use * number_blocks),
+                                    sparse=False)
+                q_1 = matmul(qsaux.pop(0), small_q)
                 if len(qsaux) > 0:
-                    q_2 = matmul(qsaux.pop(0), small_q[shape_to_use:])
+                    q_2 = matmul(qsaux.pop(0), small_q_2)
                     qs.append(concat_rows(q_1, q_2))
                 else:
-                    q_1._shape = (q_1.shape[0], q_1.shape[1] - a._reg_shape[0])
+                    q_1._shape = (q_1.shape[0], q_1.shape[1])
                     qs.append(q_1)
-                rs.append(r)
+                rs.append(r_blocks)
+                last_block_irregular = False
+            depth = depth + 1
         q = qs[0]
-        if indexes is not None:
-            q = q[:, indexes]
-        elif a._n_blocks[0] % n_reduction != 0:
-            q._shape = (q._shape[0], q._shape[1] - a._reg_shape[0])
-
-        r_blocks = [[object()]]
-        _construct_blocks(r_blocks, r, (a.shape[0], a.shape[1]))
-        r = _empty_array(shape=(a.shape[0], a.shape[1]),
-                         block_size=(a.shape[0], a.shape[1]))
-        r._blocks = r_blocks
+        r = Array(r_blocks, top_left_shape=a._reg_shape,
+                  reg_shape=a._reg_shape,
+                  shape=a.shape,
+                  sparse=False)
         return q, r
 
     elif mode == "complete_inverse":
-        if n_reduction == 2:
-            if _is_not_power_of_two(a._n_blocks[0]):
-                raise ValueError("This mode only works if the number of "
-                                 "blocks is a direct power of the "
-                                 "reduction number")
-        else:
-            if _is_not_power(n_reduction, a._n_blocks[0]):
-                raise ValueError(
-                    "This mode only works if the number of blocks is a"
-                    " direct power of the reduction number")
-
-        for block in a._blocks:
-            q, r = _compute_qr([block], "complete")
-            qs.append(q)
-            rs.append(r)
+        if _is_not_power_of_two(a._n_blocks[0]):
+            raise ValueError("This mode only works if the number of "
+                             "blocks is a direct power of the "
+                             "reduction number")
```

Comment:
The message should be: This mode only works if the number of blocks is a direct power of 2